File: notes/html.go
package notes

import (
	"fmt"
	"html/template"
	"sort"

	"github.com/karabatov/ddpub/config"
	"github.com/karabatov/ddpub/dd"
	"github.com/karabatov/ddpub/l10n"
	"github.com/karabatov/ddpub/layout"
)

func htmlForPage(note *noteContent, s *Store) (template.HTML, error) {
	cp := layout.ContentPage{
		Title:   note.title,
		Content: template.HTML(note.content),
	}
	rendered, err := layout.FillContentPage(cp)
	if err != nil {
		return "", err
	}
	return rendered, nil
}

func feedNotesListItems(t dd.Tag, w *config.WebsiteLang, s *Store) []layout.NoteListItem {
	notes := []layout.NoteListItem{}
	for _, n := range s.notesForTag(w, t) {
		nli := layout.NoteListItem{
			ListItem: layout.ListItem{
				Title: n.title,
				URL:   template.HTML(w.URLForFeedNote(n.slug)),
			},
			Date: dateForNote(&n, w),
		}
		notes = append(notes, nli)
	}
	return notes
}

func tagsListItems(w *config.WebsiteLang, s *Store) []layout.TagListItem {
	tags := []layout.TagListItem{}
	for _, t := range w.Tags {
		li := layout.TagListItem{
			ListItem: layout.ListItem{
				Title: template.HTML(t.Title),
				URL:   template.HTML(w.URLForTag(t)),
			},
			Count: len(feedNotesListItems(t.Tag, w, s)),
		}
		tags = append(tags, li)
	}
	sort.Slice(tags, func(i, j int) bool {
		return tags[i].Count > tags[j].Count
	})
	return tags
}

func htmlForBuiltinFeed(w *config.WebsiteLang, s *Store) (template.HTML, error) {
	var content template.HTML
	if len(w.Feed.ID) > 0 {
		note := s.noteContent[w.Feed.ID]
		content = template.HTML(note.content)
	}

	p := layout.BuiltinFeed{
		Title:   template.HTML(w.Feed.Title),
		Content: content,
		Notes:   feedNotesListItems(w.Feed.Tag, w, s),
	}
	rendered, err := layout.FillBuiltinFeed(p)
	if err != nil {
		return "", err
	}
	return rendered, nil
}

func htmlForBuiltinTags(w *config.WebsiteLang, s *Store) (template.HTML, error) {
	p := layout.BuiltinTags{
		Title: template.HTML(w.Str(l10n.TagsTitle)),
		Tags:  tagsListItems(w, s),
	}
	rendered, err := layout.FillBuiltinTags(p)
	if err != nil {
		return "", err
	}
	return rendered, nil
}

func htmlForTag(t *config.Tag, w *config.WebsiteLang, s *Store) (template.HTML, error) {
	var content template.HTML
	if len(t.ID) > 0 {
		note := s.noteContent[t.ID]
		content = template.HTML(note.content)
	}

	tp := layout.ContentTagPage{
		Title:   template.HTML(t.Title),
		Content: content,
		Notes:   feedNotesListItems(t.Tag, w, s),
	}
	rendered, err := layout.FillContentTagPage(tp)
	if err != nil {
		return "", err
	}
	return rendered, nil
}

func htmlForNote(note *noteContent, w *config.WebsiteLang) (template.HTML, error) {
	tags := []layout.ListItem{}
	for _, t := range w.TagsToPublished(note.tags) {
		tags = append(tags, layout.ListItem{
			Title: template.HTML(t.Title),
			URL:   template.HTML(w.URLForTag(t)),
		})
	}
	cn := layout.ContentNote{
		Title:   note.title,
		Date:    dateForNote(note, w),
		Content: template.HTML(note.content),
		Tags:    tags,
		Suffix:  template.HTML(w.NoteSuffix),
	}
	rendered, err := layout.FillContentNote(cn)
	if err != nil {
		return "", err
	}
	return rendered, nil
}

func dateForNote(note *noteContent, w *config.WebsiteLang) string {
	datePub := note.date.Format(w.Str(l10n.DateFormat))
	if note.updatedDate != note.date {
		dateUpd := note.updatedDate.Format(w.Str(l10n.DateFormat))
		return fmt.Sprintf(w.Str(l10n.DateUpdatedPublished), dateUpd, datePub)
	}

	return fmt.Sprintf(w.Str(l10n.DatePublished), datePub)
}

File: src/notes/html.rs
//! HTML content generation functions.

use crate::config::WebsiteLang;
use crate::error::{Error, Result};
use crate::l10n::Key;
use crate::layout;
use crate::notes::{NoteContent, Store};

pub fn html_for_page(note: &NoteContent) -> Result<String> {
    layout::fill_content_page(&layout::ContentPage {
        title: note.meta.title.clone(),
        content: note.content.clone(),
        has_leading_h1: note.meta.has_leading_h1,
    })
}

pub fn feed_notes_list_items(
    tag: &str,
    w: &WebsiteLang,
    s: &Store,
) -> Vec<layout::NoteListItem> {
    s.notes_for_tag(w, tag)
        .iter()
        .map(|n| layout::NoteListItem {
            list_item: layout::ListItem {
                title: n.meta.title.clone(),
                url: w.url_for_feed_note(&n.meta.slug),
            },
            date: date_for_note(n, w),
        })
        .collect()
}

pub fn tags_list_items(w: &WebsiteLang, s: &Store) -> Vec<layout::TagListItem> {
    let mut tags: Vec<layout::TagListItem> = w
        .tags
        .values()
        .map(|t| {
            let count = feed_notes_list_items(&t.tag, w, s).len();
            layout::TagListItem {
                list_item: layout::ListItem {
                    title: t.title.clone(),
                    url: w.url_for_tag(t),
                },
                count,
            }
        })
        .collect();

    tags.sort_by(|a, b| b.count.cmp(&a.count));
    tags
}

pub fn html_for_builtin_feed(w: &WebsiteLang, s: &Store) -> Result<String> {
    let content = if !w.feed.id.is_empty() {
        s.note_content
            .get(&w.feed.id)
            .map(|n| n.content.clone())
            .ok_or_else(|| Error::NoteContentNotFound { id: w.feed.id.clone() })?
    } else {
        String::new()
    };

    layout::fill_builtin_feed(&layout::BuiltinFeed {
        title: w.feed.title.clone(),
        content,
        notes: feed_notes_list_items(&w.feed.tag, w, s),
    })
}

pub fn html_for_builtin_tags(w: &WebsiteLang, s: &Store) -> Result<String> {
    layout::fill_builtin_tags(&layout::BuiltinTags {
        title: w.str(Key::TagsTitle).to_string(),
        tags: tags_list_items(w, s),
    })
}

pub fn html_for_tag(
    t: &crate::config::tag::TagConfig,
    w: &WebsiteLang,
    s: &Store,
) -> Result<String> {
    let (content, has_leading_h1) = if !t.id.is_empty() {
        let note = s.note_content
            .get(&t.id)
            .ok_or_else(|| Error::NoteContentNotFound { id: t.id.clone() })?;
        (note.content.clone(), note.meta.has_leading_h1)
    } else {
        (String::new(), false)
    };

    layout::fill_content_tag(&layout::ContentTagPage {
        title: t.title.clone(),
        content,
        notes: feed_notes_list_items(&t.tag, w, s),
        has_leading_h1,
    })
}

pub fn html_for_note(note: &NoteContent, w: &WebsiteLang) -> Result<String> {
    let tags: Vec<layout::ListItem> = w
        .tags_to_published(&note.meta.tags)
        .iter()
        .map(|t| layout::ListItem {
            title: t.title.clone(),
            url: w.url_for_tag(t),
        })
        .collect();

    layout::fill_content_note(&layout::ContentNote {
        title: note.meta.title.clone(),
        date: date_for_note(note, w),
        content: note.content.clone(),
        tags,
        suffix: w.note_suffix.clone(),
        has_leading_h1: note.meta.has_leading_h1,
    })
}

pub fn date_for_note(note: &NoteContent, w: &WebsiteLang) -> String {
    let date_format = w.str(Key::DateFormat);
    let date_pub = note.meta.date.format(date_format).to_string();

    if note.meta.updated_date != note.meta.date {
        let date_upd = note.meta.updated_date.format(date_format).to_string();
        w.str(Key::DateUpdatedPublished)
            .replacen("%s", &date_upd, 1)
            .replacen("%s", &date_pub, 1)
    } else {
        w.str(Key::DatePublished).replacen("%s", &date_pub, 1)
    }
}
